### PR TITLE
enhance: Optimeize clustering compaction

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #32939

### CODE REVIEW

File: internal/datanode/compaction/clustering_compactor.go [hunk 1]
```diff
@@ -94,13 +94,16 @@ type clusteringCompactionTask struct {
 	// vector
 	segmentIDOffsetMapping map[int64]string
 	offsetToBufferFunc     func(int64, []uint32) *ClusterBuffer
+
+	cm storage.ChunkManager
```
Comment (wayblink) [inline internal/datanode/compaction/clustering_compactor.go:98]: seems not used now

File: internal/datanode/compaction/clustering_compactor.go [hunk 2]
```diff
@@ -112,14 +115,17 @@ type ClusterBuffer struct {
 }
 
 type SpillSignal struct {
```
Comment (wayblink) [inline internal/datanode/compaction/clustering_compactor.go:117]: Let‘s rename it to FlushSignal, I miss it in the last PR

File: internal/datanode/compaction/clustering_compactor.go [hunk 3]
```diff
@@ -638,13 +662,11 @@ func (t *clusteringCompactionTask) backgroundFlush(ctx context.Context) {
...
 			var err error
-			if signal.buffer == nil {
+			if signal.writer == nil {
 				err = t.flushLargestBuffers(ctx)
 			} else {
 				err = func() error {
-					t.clusterBufferLocks.Lock(signal.buffer.id)
-					defer t.clusterBufferLocks.Unlock(signal.buffer.id)
-					return t.flushBinlog(ctx, signal.buffer)
+					return t.flushBinlog(ctx, t.clusterBuffers[signal.id], signal.writer, signal.pack)
```
Comment (wayblink) [inline internal/datanode/compaction/clustering_compactor.go:669]: submit to pool as well?

File: internal/datanode/compaction/clustering_compactor.go [hunk 4]
```diff
@@ -694,13 +743,10 @@ func (t *clusteringCompactionTask) flushAll(ctx context.Context) error {
...
 		err := func() error {
 			t.clusterBufferLocks.Lock(buffer.id)
-			defer t.clusterBufferLocks.Unlock(buffer.id)
-			err := t.flushBinlog(ctx, buffer)
-			if err != nil {
-				log.Error("flushBinlog fail")
-				return err
-			}
-			err = t.packBufferToSegment(ctx, buffer)
+			writer := buffer.writer
```
Comment (wayblink) [inline internal/datanode/compaction/clustering_compactor.go:746]: submit to pool as well？

File: internal/datanode/compaction/clustering_compactor.go [hunk 5]
```diff
@@ -1013,16 +1068,28 @@ func (t *clusteringCompactionTask) scalarPlan(dict map[interface{}]int64) [][]in
 	return buckets
 }
 
-func (t *clusteringCompactionTask) refreshBufferWriter(buffer *ClusterBuffer) error {
-	segmentID, err := t.allocator.AllocOne()
-	if err != nil {
-		return err
+func (t *clusteringCompactionTask) refreshBufferWriter(buffer *ClusterBuffer) (bool, error) {
+	var segmentID int64
```
Comment (wayblink) [inline internal/datanode/compaction/clustering_compactor.go:1072]: Can we spilt out a new method `needPack`? seems a little mix-together now.

File: internal/datanode/compaction/clustering_compactor.go (hunk 1 above)
Comment (xiaocai2333) [reply internal/datanode/compaction/clustering_compactor.go:98]: yeah, will remove

File: internal/datanode/compaction/clustering_compactor.go (hunk 3 above)
Comment (xiaocai2333) [reply internal/datanode/compaction/clustering_compactor.go:669]: ok

File: internal/datanode/compaction/clustering_compactor.go (hunk 4 above)
Comment (xiaocai2333) [reply internal/datanode/compaction/clustering_compactor.go:746]: ok

File: internal/datanode/compaction/clustering_compactor.go (hunk 5 above)
Comment (xiaocai2333) [reply internal/datanode/compaction/clustering_compactor.go:1072]: ok

**Pull request closed without merging**